src/prefect/task_engine.py

```diff
@@ -640,6 +645,19 @@ def start(
     def transaction_context(self) -> Generator[Transaction, None, None]:
         result_factory = getattr(TaskRunContext.get(), "result_factory", None)
 
+        def __prefect_task_rollback_handler(txn: Transaction):
+            # if there are no user-defined rollback hooks to run, then don't
+            # explicitly mark as rolled back
+            if txn.on_rollback_hooks == [__prefect_task_rollback_handler]:
+                return
+
+            self.set_state(
+                Completed(
+                    name="RolledBack",
+                    message="Task rolled back as part of transaction",
+                )
+            )
```

Comment:
Yup agreed fully - I think this specific state transition would benefit greatly from @chrisguidry and jake's new state-change mechanism, because in that new world all that's needed is an event emission from the transaction which is super lightweight.  In fact, I think we should only implement this transition using that new mechanic.

Reply:
Sounds good! I had the thought to wait for some of the client-side orchestration to hit main to simplify the state transitioning. Now that's in main and also:

> I think we should only implement this transition using that new mechanic.

I'm gonna go look into this 👍